### ISSUE TITLE
Question: cardinality of dice loss seems constant

### ISSUE THREAD

**Issue by keven425** [kornia/kornia#430]
This is likely a general question with dice loss itself, not with the library.

cardinality = torch.sum(input_soft + target_one_hot, dims)

The cardinality seems constant, regardless of predictions. This is because, :

both input_soft and target_one_hot are of shape (N, C, H, W) where C = number of classes.

torch.sum(input_soft, dims) is constant, as every C dimension's softmax probability adds up to 1.
torch.sum(target_one_hot, dims) is constant, as every C dimension's one hot adds up to C
therefore, the sum, cardinality is also constant
Although the nominator (intersection) changes according to prediction, the denominator (cardinality) does not. Is this valid?

**Comment by indigoviolet** [comment]
Yeah, I agree with this. 

Further, the `intersection` is `torch.sum(input_soft * target_one_hot, dims)`, which works out to `sum(prob_of_correct_class, dims)` (because all the other products are zero). 

As described above, `cardinality` will always be `2 * H * W`

`dice_loss` then becomes, for a single image: `1 - 2 * sum(prob_of_correct_class)/2 * H * W`, which is minimized when the prob of correct class is zero?

I think the problem is that you can't sum over the classes in intersection and cardinality, but have to do so after taking their ratio.

Something like:

```
intersection = torch.sum(input_soft * target_one_hot, (1,2))
cardinality = torch.sum(input_soft + target_one_hot, (1,2))
dice_loss = 1 - torch.sum(2 * intersection/cardinality, (0)).mean()
```

Also see: https://github.com/kevinzakka/pytorch-goodies/blob/master/losses.py#L54
